Add scopes to metrics explore

bfmatei opened the pull request:
**What is this feature?**

This PR brings support for scopes in the metrics explore area.

What it actually contains:
- make use of `ScopesFacade` behavior to:
   - show the scopes selector
   - subscribe to scopes changes so it refreshes the queries
- make use of suggestions API as opposed to the labels and label values one to get the labels and label values for the adhoc filters
   - this is particularly interesting because right now it makes use of what queries there are currently and it returns the suggestions only for those (as well as respecting the scopes and other adhoc filters)

**Why do we need this feature?**

Ease the usage of metrics explore.

**Who is this feature for?**

-

**Which issue(s) does this PR fix?**:

-

**Special notes for your reviewer:**

I left a couple of comments to explain better why something was added.

Please note that the OTel use case was not handled (left some TODOs) since I was not able to test this scenario.

There are a couple of feature toggles that need to exist so this works fine. I haven't tested this PR w/o the feature toggles.

```
[feature_toggles]
; enables scenes as scopes make use of those
scenes = true
; enable scopes
scopeFilters = true
; enables the K8s API server since scopes are built on top of it
grafanaAPIServer = true
grafanaAPIServerWithExperimentalAPIs = true
; enable passing scopes to Prometheus queries and other Prometheus APIs
promQLScope = true
; enable integration of scopes in metrics explore
enableScopesInMetricsExplore = true
```

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

bfmatei committed:
Scopes: Use scopes in metrics explore

bfmatei committed:
Fix for passing scopes to queries

bfmatei committed:
Just remove it.

bfmatei committed:
Revert some stuff

bfmatei committed:
Add queries

bfmatei committed:
Add comment for suggestions API

itsmylife commented on public/app/features/trails/MetricSelect/MetricSelectScene.tsx:
```diff
@@ -220,45 +222,49 @@ export class MetricSelectScene extends SceneObjectBase<MetricSelectSceneState> i
-        noOtelMetrics = true;
-      }
-    }
+    // TODO: fix this
+    // if (trail.state.useOtelExperience) {
+    //   const jobsList = trail.state.otelTargets?.jobs;
+    //   const instancesList = trail.state.otelTargets?.instances;
+    //   // no targets have this combination of filters so there are no metrics that can be joined
+    //   // show no metrics
+    //   if (jobsList && jobsList.length > 0 && instancesList && instancesList.length > 0) {
+    //     const otelMatches = limitOtelMatchTerms(matchTerms, jobsList, instancesList, missingOtelTargets);
+    //
+    //     missingOtelTargets = otelMatches.missingOtelTargets;
+    //
+    //     matchTerms.push(otelMatches.jobsRegex);
+    //     matchTerms.push(otelMatches.instancesRegex);
+    //   } else {
+    //     noOtelMetrics = true;
+    //   }
+    // }
```
@bohandley You might want to check this.

bfmatei replied:
Pushed update to fix  this.

itsmylife commented on public/app/features/trails/MetricSelect/MetricSelectScene.tsx:
```diff
@@ -28,13 +28,15 @@ import {
 import { getMetricDescription } from '../helpers/MetricDatasourceHelper';
 import { reportExploreMetrics } from '../interactions';
-import { limitOtelMatchTerms } from '../otel/util';
+// TODO: fix this
+// import { limitOtelMatchTerms } from '../otel/util';
```
@bfmatei Do you need support for this fix or just commented out temporarily?

bfmatei replied:
They were commented out temporarily as I couldn't test it. I'm revisiting this.

bfmatei replied:
Pushed update to fix it.

bfmatei committed:
Add feature toggle

bfmatei committed:
Push

bfmatei commented:
Added `enableScopesInMetricsExplore` feature flag to allow metrics explore to work w/o scope (as it currently does).

I also fixed the OTel use case (I think?).

bfmatei committed:
Fix tests and linting

bergquist commented:
/deploy-to-hg

bfmatei committed:
Fix scopes selector for new nav system

bergquist committed:
Suggestion handler should always inject scope in _some_ way.  (#95444)

always pass scope when exploring metrics

Signed-off-by: bergquist <carl.bergquist@gmail.com>

bfmatei commented:
/deploy-to-hg

bfmatei commented:
/deploy-to-hg

bfmatei committed:
Refetch metrics instead of rerunning the queries

bfmatei commented:
/deploy-to-hg

bergquist committed:
explore: only pass queries to suggestion if the list is short (#95540)

Signed-off-by: bergquist <carl.bergquist@gmail.com>

bergquist committed:
frontend linting

Signed-off-by: bergquist <carl.bergquist@gmail.com>

bergquist committed:
linting AGAIN

Signed-off-by: bergquist <carl.bergquist@gmail.com>

bergquist commented:
/deploy-to-hg

bfmatei committed:
Fix for metrics refresh

itsmylife commented on public/app/features/trails/utils.ts:
```diff
@@ -175,12 +205,73 @@ export function limitAdhocProviders(
+): Promise<FetchResponse<SuggestionsResponse>> {
+  return await lastValueFrom(
+    getBackendSrv().fetch<SuggestionsResponse>({
+      url: `/api/datasources/uid/${dataSourceUid}/resources/suggestions`,
```
Is there a case where we don't have suggestion API not available? If so, should we handle the errors?

itsmylife reviewed (commented):
The code looks good. I added some nitpicks. 
I'd like to test this on the browser and I failed to create the right environment.
After the test, I'll approve.
Great job 🚀

itsmylife commented on public/app/features/trails/DataTrailsApp.tsx:
```diff
@@ -81,11 +125,35 @@ let dataTrailsApp: DataTrailsApp;
+              const trail = facade.parent && 'trail' in facade.parent.state ? facade.parent.state.trail : undefined;
+
+              if (trail instanceof DataTrail) {
+                trail.publishEvent(new RefreshMetricsEvent());
```
```suggestion
                trail.publishEvent(new RefreshMetricsEvent());
                trail.checkDataSourceForOTelResources();
```

itsmylife reviewed (approved):
Great job 🎉

bohandley reviewed (approved):
LGTM! Thank you for working on this @bfmatei 🚀

bfmatei committed:
Update public/app/features/trails/DataTrailsApp.tsx

Co-authored-by: ismail simsek <ismailsimsek09@gmail.com>

bfmatei committed:
Fix tests
